ansible/roles-infra/infra-osp-template-generate/templates/cloud_template_master.j2

```diff
@@ -14,7 +14,7 @@ resources:
     properties:
       name: {{ guid }}-infra_key
       save_private_key: true
-
```

Comment:
Can't we just add the user in the `OS::Nova::KeyPair` resource ?


https://docs.openstack.org/heat/pike/template_guide/openstack.html#OS::Nova::KeyPair-prop-user

```yaml
  {{ guid }}-infra_key:
    type: OS::Nova::KeyPair
    properties:
      name: {{ guid }}-infra_key
      user: {{ guid }}-project_user
      save_private_key: true
    depends_on:
      - {{ guid }}-project_user
```

Reply:
Huh, this is a good catch. I'm guess I missed this and it would potentially make things simpler.  I'll set up a test to make sure it is cleaned up correctly and if so I can simplify this change and keep everything as part of the tempalte.

Reply:
Ok, I remember why something like this doesn't work. This whole keypair-associated-to-user thing makes all this more complicated.

If I create the Heat stack as the admin user, I am doing it as the admin user...even though it is in the `{{ guid }}-user` project. Even though I can tell the Heat template to create the keypair AS or FOR another user, when my Heat stack goes to launch instances it does not have access to the keypair. Because I am launching the instances as admin...but I am trying to use a keypair from the user. Remember keypairs are 100% associated to user and not project or anything else.

But why don't we launch the Heat stack as the user as it is today? Chicken and egg...the user gets created with the Heat stack so I still have to create the Heat stack as admin. And then it is back to launching everything in the stack as admin as well.

Moving the user creation out of the Heat stack gives us the ability to create all the things as the user rather than admin. We just have to make sure we do the last bit of cleanup on the user account + keypair and everything else stays the same.

@fridim make sense?

Reply:
Yes it makes sense.

How urgent is this PR?  Since we're at moving things out of the heat template and changing their "owner", I'm thinking maybe we should switch gears and have 2 HOT templates:

1. One admin HOT that will create project, user, rolebindings, quotas, etc, like it's done here: https://github.com/redhat-gpe/novello-templates/blob/master/heat-templates/EMEA-PARTNERS-ansible_tower_summit-v1.0-bp/stack_admin.yaml
2. The user HEAT template.

IIUC:
- today, 1. is done by some ansible tasks (not heat) **and** the heat template. 2. is done in the heat template, as admin.  Which is confusing and wrong now that we discovered the issue with keypairs.
- with this PR, 1. is done entirely outside of the heat template as ansible tasks executed as admin user in our osp cloud-provider roles. and 2. is done in the heat template executed as user.

In the future, maybe we want to use HOT for the admin part as well since it would simplify the deletion.

Would that be a too big change to be done now ? Wdyt?

The changes in that PR are valid since they address the issue and fix them.

Reply:
It is not urgent-urgent...but it should be done, especially before the next set of OpenStack clusters get deployed and used. I also didn't want this branch to get too far out of sync with what everything was tested against.

As for your suggestion, I'm not opposed, but not in favor either. I find HOT (and CFN and ARM and...) more complicated to use and maintain. I get the need for them when we are deploying a more complex set of resources that may be dependent on one another and have nuance in how they are instantiated and removed. 

However, for doing simple things like we are talking about here, I really find Ansible tasks much easier to create and maintain. They do the job well to create these resources. Teardown of these same resources is very predictable as well.

If you feel strongly about moving these into a HOT, I will do it...but I'm also quite satisfied with them being Ansible tasks.